test/test-reuseport-unix.c

```diff
@@ -0,0 +1,149 @@
+#include <sys/un.h>
+#endif
+#ifdef EVENT__HAVE_SYS_SOCKET_H
+#include <sys/socket.h>
+#endif
+
+#include <event2/listener.h>
+
+static void
+listener_cb(struct evconnlistener *listener, evutil_socket_t fd,
+    struct sockaddr *sa, int socklen, void *user_data)
+{
+	printf("Empty listener, do nothing about it!\n");
+}
+
+int
+main(int argc, char **argv)
+{
+	struct event_base *base;
+	struct evconnlistener *listener;
+	struct sockaddr_un addr;
+
+#ifdef _WIN32
+	DWORD n;
+	TCHAR tempPath[MAX_PATH];
+	TCHAR longTempPath[MAX_PATH];
+#ifdef EVENT__HAVE_AFUNIX_H
+	if (evutil_check_working_afunix_() == 0)
+		/* AF_UNIX is not available on the current Windows platform,
+		 * just mark this test as success.
+		 */
+		return EXIT_SUCCESS;
+#endif
+	n = GetTempPathW(MAX_PATH, tempPath);
+	if (n == 0 || n < MAX_PATH)
+		return EXIT_FAILURE;
+	n = GetLongPathNameW(tempPath, longTempPath, MAX_PATH);
+	if (n == 0 || n >= MAX_PATH)
+		return EXIT_FAILURE;
+	TCHAR socket_path[MAX_PATH];
+	_stprintf(socket_path, _T("%stest-reuseport-unix.sock"), longTempPath);
+	/* For security reason, we must delete any existing sockets in the filesystem. */
+	DeleteFileW(socket_path);
+#else
+	char socket_path[] = "/tmp/test-reuseport-unix.sock";
+	/* For security reason, we must delete any existing sockets in the filesystem. */
+	unlink(socket_path);
+#endif
+
+#ifdef _WIN32
+	WSADATA wsaData;
+	int r;
+	/* Initialize Winsock. */
+	r = WSAStartup(MAKEWORD(2,2), &wsaData);
+	if (r) {
+		fprintf(stderr, "WSAStartup failed with error: %d\n", r);
+		return EXIT_FAILURE;
+	}
+#endif
+
+	base = event_base_new();
+	if (!base) {
+		fprintf(stderr, "Could not initialize libevent!\n");
+		return EXIT_FAILURE;
+	}
+
+	memset(&addr, 0, sizeof(addr));
+	addr.sun_family = AF_UNIX;
+	strncpy(addr.sun_path, socket_path, sizeof(addr.sun_path)-1);
+
+	listener = evconnlistener_new_bind(base, listener_cb, (void *)base,
+	    LEV_OPT_REUSEABLE|LEV_OPT_CLOSE_ON_FREE, -1,
+			(struct sockaddr*)&addr, sizeof(addr));
+
+	if (listener) {
+		fprintf(stderr, "AF_UNIX listener shouldn't use SO_REUSEADDR!\n");
+		evconnlistener_free(listener);
+		return EXIT_FAILURE;
+	}
+
+	listener = evconnlistener_new_bind(base, listener_cb, (void *)base,
+	    LEV_OPT_REUSEABLE_PORT|LEV_OPT_CLOSE_ON_FREE, -1,
+			(struct sockaddr*)&addr, sizeof(addr));
+
+	if (listener) {
+		fprintf(stderr, "AF_UNIX listener shouldn't use SO_REUSEPORT!\n");
+		evconnlistener_free(listener);
+		return EXIT_FAILURE;
+	}
+
+	/* Create a AF_UNIX listener without reusing address or port. */
+	listener = evconnlistener_new_bind(base, listener_cb, (void *)base,
+	    LEV_OPT_CLOSE_ON_EXEC|LEV_OPT_CLOSE_ON_FREE, -1,
+			(struct sockaddr*)&addr, sizeof(addr));
+	if (!listener) {
+		fprintf(stderr, "Could not create a AF_UNIX listener!\n");
+		return EXIT_FAILURE;
+	}
+
+	evconnlistener_free(listener);
+	event_base_free(base);
+	printf("Tested successfully!\n");
+
+#ifdef _WIN32
+	WSACleanup();
+#endif
+
+	return EXIT_SUCCESS;
```

Comment:
Thank you for the test! Why did you prefer to write a separate test file instead of integrating it to `test/regress_listener.c`?

Reply:
No particular reason, I can move it to `regress_listener.c` if you think that makes more sense.

Reply:
Yes please

Reply:
Done.